pkg/restore/restore.go

```diff
@@ -259,50 +259,98 @@ func (ctx *context) execute() (api.RestoreResult, api.RestoreResult) {
 
-	// namespace-scoped
-	namespacesPath := path.Join(dir, api.NamespaceScopedDir)
-	exists, err = ctx.fileSystem.DirExists(namespacesPath)
+	// Make sure the top level "resources" dir exists:
+	resourcesDir := path.Join(dir, api.ResourcesDir)
+	_, err := ctx.fileSystem.DirExists(resourcesDir)
 	if err != nil {
 		addArkError(&errors, err)
 		return warnings, errors
 	}
-	if !exists {
-		return warnings, errors
-	}
 
-	nses, err := ctx.fileSystem.ReadDir(namespacesPath)
+	resourceDirs, err := ctx.fileSystem.ReadDir(resourcesDir)
 	if err != nil {
 		addArkError(&errors, err)
 		return warnings, errors
 	}
 
-	namespaceFilter := collections.NewIncludesExcludes().Includes(ctx.restore.Spec.IncludedNamespaces...).Excludes(ctx.restore.Spec.ExcludedNamespaces...)
-	for _, ns := range nses {
-		if !ns.IsDir() {
+	resourceDirsMap := make(map[string]os.FileInfo)
+
+	for _, rscDir := range resourceDirs {
+		rscName := rscDir.Name()
+		resourceDirsMap[rscName] = rscDir
+	}
+
+	for _, resource := range ctx.prioritizedResources {
+		rscDir := resourceDirsMap[resource.String()]
+		if rscDir == nil {
 			continue
 		}
-		nsPath := path.Join(namespacesPath, ns.Name())
 
-		if !namespaceFilter.ShouldInclude(ns.Name()) {
-			ctx.infof("Skipping namespace %s", ns.Name())
-			continue
+		resourcePath := path.Join(resourcesDir, rscDir.Name())
+
+		clusterSubDir := path.Join(resourcePath, api.ClusterScopedDir)
+		clusterSubDirExists, err := ctx.fileSystem.DirExists(clusterSubDir)
+		if err != nil {
+			addArkError(&errors, err)
+			return warnings, errors
+		}
+		if clusterSubDirExists {
+			w, e := ctx.restoreResource(resource.String(), "", clusterSubDir)
+			merge(&warnings, &w)
+			merge(&errors, &e)
+		}
+
+		nsSubDir := path.Join(resourcePath, api.NamespaceScopedDir)
+		nsSubDirExists, err := ctx.fileSystem.DirExists(nsSubDir)
+		if err != nil {
+			addArkError(&errors, err)
+			return warnings, errors
 		}
+		if nsSubDirExists {
+			nsDirs, err := ctx.fileSystem.ReadDir(nsSubDir)
+			if err != nil {
+				addArkError(&errors, err)
+				return warnings, errors
+			}
+
+			for _, nsDir := range nsDirs {
+				if !nsDir.IsDir() {
+					continue
+				}
+				nsPath := path.Join(nsSubDir, nsDir.Name())
+
+				if !namespaceFilter.ShouldInclude(nsDir.Name()) {
+					ctx.infof("Skipping namespace %s", nsDir.Name())
+					continue
+				}
 
-		w, e := ctx.restoreNamespace(ns.Name(), nsPath)
-		merge(&warnings, &w)
-		merge(&errors, &e)
+				nsName := nsDir.Name()
+
+				// fetch mapped NS name
+				// TODO: what is a mapped ns name? for renames?
+				mappedNsName := nsName
+				if target, ok := ctx.restore.Spec.NamespaceMapping[nsName]; ok {
+					mappedNsName = target
+				}
+
+				// ensure namespace exists
+				ns := &v1.Namespace{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: mappedNsName,
+					},
+				}
+				if _, err := kube.EnsureNamespaceExists(ns, ctx.namespaceClient); err != nil {
+					addArkError(&errors, err)
+					return warnings, errors
```

Comment:
Shouldn't this continue the `nsDirs` loop on line 318 to try additional namespaces?

Reply:
Whoah, yeah that's a bad artifact from the refactor, good catch. Changing to continue.